lib/instrumentation/undici.js

```diff
@@ -0,0 +1,254 @@
+      'diagnostics_channel or feature_flag.undici_instrumentation = false. Skipping undici instrumentation.'
+    )
+    return
+  }
+
+  /**
+   * This event occurs after the Undici Request is created
+   * We will check current segment for opaque and also attach
+   * relevant headers to outgoing http request
+   *
+   * @param {Object} params
+   * @param {Object} params.request undici request object
+   */
+  diagnosticsChannel.channel('undici:request:create').subscribe(({ request }) => {
+    const parent = agent.tracer.getSegment()
+    request[SYMBOLS.PARENT_SEGMENT] = parent
+    if (!parent || (parent && parent.opaque)) {
+      logger.trace(
+        'Not capturing data for outbound request (%s) because parent segment opaque (%s)',
+        request.path,
+        parent.name
+      )
+
+      return
+    }
+
+    const transaction = parent.transaction
+    const outboundHeaders = Object.create(null)
+    if (agent.config.encoding_key && transaction.syntheticsHeader) {
+      outboundHeaders[NEWRELIC_SYNTHETICS_HEADER] = transaction.syntheticsHeader
+    }
+
+    if (agent.config.distributed_tracing.enabled) {
+      transaction.insertDistributedTraceHeaders(outboundHeaders)
+    } else if (agent.config.cross_application_tracer.enabled) {
+      cat.addCatHeaders(agent.config, transaction, outboundHeaders)
+    } else {
+      logger.trace('CAT disabled, not adding headers!')
+    }
+
+    // eslint-disable-next-line guard-for-in
+    for (const key in outboundHeaders) {
+      request.addHeader(key, outboundHeaders[key])
+    }
+  })
+
+  /**
+   * This event occurs right before the data is written to the socket.
+   * Undici has some abstracted headers that are only created at this time, one
+   * is the `host` header which we need to name the Undici segment. So in this
+   * handler we create, start and set the segment active, name it, and
+   * attach the url/procedure/request.parameters
+   *
+   * @param {Object} params
+   * @param {Object} params.request undicie request object
+   * @param {TLSSocket | net.Socket} socket active socket connection
+   *
+   */
+  diagnosticsChannel.channel('undici:client:sendHeaders').subscribe(({ request, socket }) => {
+    const parentSegment = request[SYMBOLS.PARENT_SEGMENT]
+    if (!parentSegment || (parentSegment && parentSegment.opaque)) {
+      return
+    }
+
+    const port = socket.remotePort
+    const isHttps = socket.servername
+    let urlString
+    if (isHttps) {
+      urlString = `https://${socket.servername}`
+      urlString += port === 443 ? request.path : `:${port}${request.path}`
+    } else {
+      urlString = `http://${socket._host}`
+      urlString += port === 80 ? request.path : `:${port}${request.path}`
+    }
+
+    const url = new URL(urlString)
+
+    const name = NAMES.EXTERNAL.PREFIX + url.host + url.pathname
+    const segment = agent.tracer.createSegment(
+      name,
+      recordExternal(url.host, 'undici'),
+      request[SYMBOLS.PARENT_SEGMENT]
+    )
```

Comment:
this was copied from http-outbound. they are accessing the same thing right? i can def update but why does http-outbound access it via agent.tracer?

Reply:
Because http-outbound has not yet been converted to use the shim API. Same for the standard http instrumentation.

They are accessing the same thing but ideally, if you go through a proper instrumentation API, some things can change behind the scenes without breaking.

Really ideally, I'd love to get away from just about anything have a handle on the 'agent' itself and potentially the tracer too (other than special cases). The agent is in extreme want of more passing of dependencies vs the agent to do everything. 

In the end, this should work fine but I don't love mixing and matching usage of the 'shim' and the 'tracer'. The usage of shimmer and direct usage of 'tracer' pre-date the attempt at a instrumentation api in the shim.

Reply:
*obviously the shim has many limitations and doesn't seem to support all the needed use cases yet. i'd also like to use apollo and this and maybe fastify as it evolves to build a strong map for how we might evolve the 'shim' api into something that better supports using just what you need and composing functionality. and you know, maybe we actually have something called 'instrumentationApi' or some such and not a 'shim' :P